File: benches/three/mandelbrot.c
/* The same algorithm as `mandelbrot.ls`, in C.
 *
 * CHECKED=1 traps on integer overflow the way lex-sys does, so the two
 * halves bracket what the guarantee costs a mature backend.
 *
 * Note what this file relies on that C does not promise: `>>` on a
 * negative signed value is *implementation-defined* in C, and this code
 * needs it to be arithmetic. gcc and clang both do that; the standard
 * does not say they must. lex-sys defines it (`docs/bitwise.md` §2). */
#include <stdio.h>

#define LIMIT (4 << 16)

#if CHECKED
#include <stdlib.h>
static long mul(long a, long b) {
    long r;
    if (__builtin_smull_overflow(a, b, &r)) __builtin_trap();
    return r;
}
static long add(long a, long b) {
    long r;
    if (__builtin_saddl_overflow(a, b, &r)) __builtin_trap();
    return r;
}
static long sub(long a, long b) {
    long r;
    if (__builtin_ssubl_overflow(a, b, &r)) __builtin_trap();
    return r;
}
#else
static long mul(long a, long b) { return (long)((unsigned long)a * (unsigned long)b); }
static long add(long a, long b) { return (long)((unsigned long)a + (unsigned long)b); }
static long sub(long a, long b) { return (long)((unsigned long)a - (unsigned long)b); }
#endif

__attribute__((noinline))
static long escape(long cx, long cy, long maxiter) {
    long zx = 0, zy = 0, zx2 = 0, zy2 = 0, i = 0;
    while (i < maxiter && add(zx2, zy2) <= LIMIT) {
        long zxy = mul(zx, zy) >> 16;
        zx = add(sub(zx2, zy2), cx);
        zy = add(mul(2, zxy), cy);
        zx2 = mul(zx, zx) >> 16;
        zy2 = mul(zy, zy) >> 16;
        i = add(i, 1);
    }
    return i;
}

__attribute__((noinline))
static long grid(long width, long height, long maxiter) {
    long total = 0;
    for (long py = 0; py < height; py++) {
        long cy = ((py * 163840) / height) - 81920;
        for (long px = 0; px < width; px++) {
            long cx = ((px * 163840) / width) - 131072;
            total = add(total, escape(cx, cy, maxiter));
        }
    }
    return total;
}

int main(void) {
    printf("%ld\n", grid(400, 400, 1000));
    return 0;
}

File: benches/three/mandelbrot.rs
// The same algorithm as `mandelbrot.ls`, in Rust.
//
// Built twice: `-C overflow-checks=on` matches lex-sys's semantics, and
// off is Rust's own release default, which wraps. The gap between those
// two rows is the same gap `docs/overflow-cost.md` measured, paid by a
// different compiler.
//
// `>>` on a signed integer is arithmetic in Rust, which is defined
// rather than implementation-defined as it is in C.

const LIMIT: i64 = 4 << 16;

#[inline(never)]
fn escape(cx: i64, cy: i64, maxiter: i64) -> i64 {
    let (mut zx, mut zy, mut zx2, mut zy2, mut i) = (0i64, 0i64, 0i64, 0i64, 0i64);
    while i < maxiter && zx2 + zy2 <= LIMIT {
        let zxy = (zx * zy) >> 16;
        zx = zx2 - zy2 + cx;
        zy = 2 * zxy + cy;
        zx2 = (zx * zx) >> 16;
        zy2 = (zy * zy) >> 16;
        i += 1;
    }
    i
}

#[inline(never)]
fn grid(width: i64, height: i64, maxiter: i64) -> i64 {
    let mut total = 0i64;
    for py in 0..height {
        let cy = ((py * 163840) / height) - 81920;
        for px in 0..width {
            let cx = ((px * 163840) / width) - 131072;
            total += escape(cx, cy, maxiter);
        }
    }
    total
}

fn main() {
    println!("{}", grid(400, 400, 1000));
}
